protocol/packages/currencies/build/main.rs

```diff
@@ -0,0 +1,148 @@
+use std::{
+    env,
+    fs::File,
+    io::{self, Write},
+    path::{Path, PathBuf},
+};
+
+use anyhow::{bail, Context as _, Result};
+
+use topology::{CurrencyDefinitions, Topology};
+
+use self::{currencies_tree::CurrenciesTree, either::Either, protocol::Protocol};
+
+mod convert_case;
+mod currencies_tree;
+mod either;
+mod protocol;
+mod sources;
+mod subtype_lifetime;
+
+const PROTOCOL_JSON: &str = "../../../build-configuration/protocol.json";
+
+const TOPOLOGY_JSON: &str = "../../../build-configuration/topology.json";
+
+const BUILD_REPORT: &str = "CURRENCIES_BUILD_REPORT";
+
+fn main() -> Result<()> {
+    for path in ["build/", PROTOCOL_JSON, TOPOLOGY_JSON] {
+        println!("cargo::rerun-if-changed={path}");
+    }
+
+    println!("cargo::rerun-if-env-changed={BUILD_REPORT}");
+
+    let output_directory: &Path = &PathBuf::from(
+        env::var_os("OUT_DIR").context("Cargo did not set `OUT_DIR` environment variable!")?,
+    );
+
+    let files_exist = IntoIterator::into_iter([PROTOCOL_JSON, TOPOLOGY_JSON])
+        .map(Path::new)
+        .map(Path::try_exists)
+        .try_fold(true, |all_exist, result| {
+            result
+                .map(|exists| all_exist && exists)
+                .context("Failed to check whether JSON descriptor file exists!")
+        })?;
+
+    let build_report = if let Some(report_file) = env::var_os(BUILD_REPORT) {
+        if files_exist {
+            Either::Left(
+                File::create(report_file).context("Failed to open build report for writing!")?,
+            )
+        } else {
+            bail!(
+                "`{BUILD_REPORT:?}` environment variable set but topology \
+                and/or protocol descriptors don't exist!",
+            );
+        }
+    } else {
+        Either::Right(io::stderr())
+    };
+
+    if files_exist {
+        generate_currencies(
+            build_report,
+            output_directory,
+            serde_json::from_reader(
+                File::open(TOPOLOGY_JSON).context("Failed to open \"topology.json\"!")?,
+            )
+            .context("Failed to parse topology JSON!")?,
+            serde_json::from_reader(
+                File::open(PROTOCOL_JSON).context("Failed to open \"protocol.json\"!")?,
+            )
+            .context("Failed to parse protocol JSON!")?,
+        )
+    } else {
+        Ok(())
```

Comment:
I am curious why we say OK if the descriptors are missing.

Reply:
This case is valid as we have the dummy currencies gated behind `testing`, which do not require any additional sources.

Reply:
may we check the presence of that gate, and Err() otherwise?